### ISSUE TITLE
How to actually use this with dropbox

### ISSUE THREAD

**Issue by Oxmaster** [Paraphraser/IOTstackBackup#2]
I'm a total linux noob. Following the instructions was quite a journey, but I think I succeded with adding the dropbox part.

I want to have a weekly or monthly backup of influxdb databases on dropbox. In https://github.com/Paraphraser/IOTstackBackup#-the-backup-side-of-things- there is nothing about doing scheduled backups.

What I think I have be doing is calling  `iotstack_backup_influxdb dbbackup2021-06-23.zip` and then `rclone copy dbbackup2021-06-23.zip dropbox:IOTStackBackups`. But how can I call the script to create a datestamp with the file and then copy it to the dropbox folder automatically?

**Comment by Paraphraser** [comment]
I am going to have to make a few assumptions:

1. That you have installed IOTstackBackup as per [its setup instructions](https://github.com/Paraphraser/IOTstackBackup#-setup-).

2. That you have a good reason for running `iotstack_backup_influxdb` by itself, rather than letting `iotstack_backup` do it for you.

	> I'm supposing something like you don't want to incur the bandwidth overheads of synching with DropBox every day.
	
3. That you don't want to keep the results of running `iotstack_backup_influxdb` in `~/IOTstack/backups` because you want to keep it separate for some reason.
4. That your use of the `.zip` extension is either a typo or a misunderstanding.

So, let's start with the best way to fire off a weekly task. You need a "cron job".

```
SHELL=/bin/bash
HOME=/home/pi
PATH=/home/pi/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin

00	07 01	*	*	my_influx_backup >>./Logs/my_influx_backup.log 2>&1
```

In words, "at 07:00 on the first day of every month, run the command `my_influx_backup` and send all its output to a log file".

If you don't know how to set everything up, you:

1. Select the text in the above and copy it to the clipboard.
2. Get it into a file on your Raspberry Pi eg `mycrontab.txt`.
3. Run the commands:

	```
	$ mkdir -p ~/Logs
	$ crontab mycrontab.txt
	```

The only caveat here is to avoid doing any of the steps on a Windows machine because that risks adding CR+LF line endings and those **will** screw things up. If you haven't already done so, there really is no substitute for learning one of the Unix text editors.

> I use `vi` but that's because my first encounter with Unix was in 1982 and I haven't bothered to learn anything newer. Lazy as! Plus, I'm a Mac user so I don't have to worry about CR+LF line endings.

Now we need to create the file at this path:

```
~/.local/bin/my_influx_backup
```

The content will look something like this:

```
#!/usr/bin/env bash

#define the folder where you want to keep your influx backups
MY_INFLUX_BACKUP_FOLDER="$HOME/my_influx_backups"

# make sure that folder exists
mkdir -p "$MY_INFLUX_BACKUP_FOLDER"

# form the filename that you want
MY_INFLUX_BACKUP_FILE="dbbackup$(date +"%Y-%m-%d").tar"

# form the path
MY_INFLUX_BACKUP="$MY_INFLUX_BACKUP_FOLDER/$MY_INFLUX_BACKUP_FILE"

# run the backup process
iotstack_backup_influxdb "$MY_INFLUX_BACKUP"

# now do what you want with the resulting file. Your choices are to:
# 1. refer to the backup file by its path name: "$MY_INFLUX_BACKUP"
# 2. change into the backup directory with: cd "$MY_INFLUX_BACKUP_FOLDER"
#    and then refer to the file by its filename: "$MY_INFLUX_BACKUP_FILE"
```

You'll have to figure out the bit at the end.

Once that file is in place, remember that it will need execute permission:

```
$ chmod +x ~/.local/bin/my_influx_backup
```

You can test it just by running:

```
$ my_influx_backup
```

All the output will be written to the console. If you want to simulate what will happen when running it as a cron jon, then just copy and paste the command from the crontab (replacing the "." with "~" if you are not in your home directory). Something like:

```
$ my_influx_backup >>~/Logs/my_influx_backup.log 2>&1
```

and then you can go and look at the log output with:

```
$ cat ~/Logs/my_influx_backup.log
```

On the topic of logs, when `iotstack_backup` is called from `cron`, the log in the Logs folder is only a log of last resort (if something goes wrong with cron). The logs for the entire run of `iotstack_backup` wind up in `~/IOTstack/backups`. They get synchronised with Dropbox **and** they get pruned.

If you do it the way I have outlined here, your log will grow forever until you decide to prune it yourself. Although it's just as easy to delete the file and let `cron` recreate it on the next run, the standard Unix way of pruning a log is:

```
$ cat /dev/null >~/Logs/my_influx_backup.log
```

<hr>

Let's get back to my 4<sup>th</sup> assumption. When you run `iotstack_backup_influxdb` it produces a tar ("tape archive") not a zip.

The `iotstack_backup_influxdb` script doesn't check that you have used the correct `.tar` extension. It doesn't matter what extension you use (.zip, .fred, .etc) the content of the file will still be a `.tar`. Although most processes rely more on signature bytes than extensions to determine file content, there is still some risk of confusing downstream processes if you don't use the correct extension.

You will note that the `my_influx_backup` script uses `.tar` rather than `.zip` but otherwise gives you the filename format you wanted.

You said you were a "linux noob" but I never know how to interpret that. At the risk of telling you things you already know (and apologising in advance if I do), I'll drill down into compression formats, starting with `.zip`.

What the zip format allows you to do is compress one or more files and folders.

The `.gz` or gzip format, on the other hand, only compresses a **single** file.

Here's a practical example of a folder containing six files:

```
$ tree test
test
├── test.1.jpg
├── test.2.jpg
├── test.3.jpg
├── test.4.jpg
├── test.5.jpg
└── test.6.jpg
```

If I use `zip` on that:

```
$ zip -r test.zip test
  adding: test/test.1.jpg (deflated 2%)
  adding: test/test.2.jpg (deflated 2%)
  adding: test/test.3.jpg (deflated 2%)
  adding: test/test.4.jpg (deflated 2%)
  adding: test/test.5.jpg (deflated 2%)
  adding: test/test.6.jpg (deflated 2%)

$ ls --sort=size -l test.*
-rw-r--r-- 1 pi pi 13826362 Jun 24 14:59 test.zip

$ tree test
test
├── test.1.jpg
├── test.2.jpg
├── test.3.jpg
├── test.4.jpg
├── test.5.jpg
└── test.6.jpg
```

So, `zip` has compressed the originals, added the compressed versions to `test.zip`, and has left the originals unchanged. Now I'll do it with `gzip`:

```
$ gzip test/*
$ tree test
test
├── test.1.jpg.gz
├── test.2.jpg.gz
├── test.3.jpg.gz
├── test.4.jpg.gz
├── test.5.jpg.gz
└── test.6.jpg.gz
```

In this case, `gzip` has compressed each file, individually, and replaced each original with its compressed version. Unlike `zip`, it hasn't given you a single file containing the compressed originals. It's given you one-for-one.

To get the same **effect** as zip, you have to bring tar ("tape archive") into the mix and use a two-step process. Seeing as the files in the `test` directory are compressed at the moment, I'll just create a `.tar` but I will use the concatenated extension `.gz.tar` to indicate the order in which the final file was created:

```
$ tar -vcf test.gz.tar -C ./test .
./
./test.3.jpg.gz
./test.2.jpg.gz
./test.5.jpg.gz
./test.6.jpg.gz
./test.1.jpg.gz
./test.4.jpg.gz

$ ls --sort=size -l test.*
-rw-r--r-- 1 pi pi 13834240 Jun 24 14:59 test.gz.tar
-rw-r--r-- 1 pi pi 13826362 Jun 24 14:59 test.zip
```

> The `-vcf` is shorthand for `-v -c -f`. The `-v` option says "be verbose" (so we can see the files as they are added), the `-c` option tells `tar` to "create" the archive, the `-f` says the next argument (in this case `test.gz.tar` is the name of the archive. The -C is the path to the directory to be compressed (the equivalent of doing a `cd test` beforehand) and the `.` just means "this directory" where "this" is the path after the `-C`.

From the file sizes, you can see that doing things in the order "compress, then tar" results in a slightly worse outcome than doing the same thing with zip.

Now let's restore the gzip-compressed files in the `test` directory to their original form:

```
$ gunzip test/*
$ tree test
test
├── test.1.jpg
├── test.2.jpg
├── test.3.jpg
├── test.4.jpg
├── test.5.jpg
└── test.6.jpg
```

This time I'll do it the other way around, first by running the `tar` to collect everything inside the test directory into a single file:

```
$ tar -vcf test.tar -C ./test .
./
./test.4.jpg
./test.1.jpg
./test.2.jpg
./test.6.jpg
./test.5.jpg
./test.3.jpg

$ ls --sort=size -l test.*
-rw-r--r-- 1 pi pi 14049280 Jun 24 15:00 test.tar
-rw-r--r-- 1 pi pi 13834240 Jun 24 14:59 test.gz.tar
-rw-r--r-- 1 pi pi 13826362 Jun 24 14:59 test.zip
```

That `test.tar` is not yet compressed so we do the second step:

```
$ gzip test.tar
$ ls --sort=size -l test.*
-rw-r--r-- 1 pi pi 13834240 Jun 24 14:59 test.gz.tar
-rw-r--r-- 1 pi pi 13826362 Jun 24 14:59 test.zip
-rw-r--r-- 1 pi pi 13825995 Jun 24 15:00 test.tar.gz
```

Like before when I ran `gzip` on the whole `test` directory, the `test.tar` has been replaced with `test.tar.gz`. You can see that the results are *similar* in size but not *identical*. This time, the `.tar.gz` gets a slightly better outcome than the `.zip`. 

The idea of creating a `.tar` to sweep a folder hierarchy together and *then* compress it is so common that `tar` can do all the work in one step, by adding the `-z` option. I will also change the output filename so we get both results:

```
$ tar -vczf test.two.tar.gz -C ./test .
./
./test.4.jpg
./test.1.jpg
./test.2.jpg
./test.6.jpg
./test.5.jpg
./test.3.jpg

$ ls --sort=size -l test.*
-rw-r--r-- 1 pi pi 13834240 Jun 24 14:59 test.gz.tar
-rw-r--r-- 1 pi pi 13826362 Jun 24 14:59 test.zip
-rw-r--r-- 1 pi pi 13825995 Jun 24 15:00 test.tar.gz
-rw-r--r-- 1 pi pi 13825986 Jun 24 15:01 test.two.tar.gz
```

Again, a slight improvement in size.

Now, why does `iotstack_backup_influxdb` produce a `.tar`? To answer that, try running this on your system:

```
$ ls -1 ~/IOTstack/backups/influxdb/db
20210624T013838Z.manifest
20210624T013838Z.meta
20210624T013838Z.s100.tar.gz
…
20210624T013838Z.s9.tar.gz
```

> On my system, that "…" represents another 844 files with `.tar.gz` extensions.

In other words, when `influx` is told to create a backup, it does the equivalent of a `tar -czf` but it does it on the basis of one shard at a time. That's why you get multiple `.tar.gz` files.

Compressing files that are already compressed is horrendously inefficient. It consumes a lot of CPU time (and sometimes even produces a **larger** result):

```
$ /usr/bin/time -f "Wallclock: %E. CPU: %P\n" sudo zip -q -r ~/mybackup.zip ~/IOTstack/backups/influxdb/db
Wallclock: 0:27.39. CPU: 99%

$ /usr/bin/time -f "Wallclock: %E. CPU: %P\n" sudo tar -cf ~/mybackup.tar -C ~/IOTstack/backups/influxdb/db .
Wallclock: 0:01.58. CPU: 98%

$ /usr/bin/time -f "Wallclock: %E. CPU: %P\n" sudo tar -czf ~/mybackup.tar.gz -C ~/IOTstack/backups/influxdb/db .
Wallclock: 0:28.15. CPU: 102%

$ ls --sort=size -l mybackup.*
-rw-r--r-- 1 root root 241940480 Jun 24 15:18 mybackup.tar
-rw-r--r-- 1 root root 241092486 Jun 24 15:18 mybackup.zip
-rw-r--r-- 1 root root 240919048 Jun 24 15:19 mybackup.tar.gz
```

The first command runs a zip. It takes a bit over 27 seconds and maxes out one core all that time. The second command just sweeps the already-compressed files into a tar. Even though it still maxes out one core, it's only for 2 seconds. The third command is the "tar then compress". A bit longer that the zip on the wallclock and, for what? An improvement in file size of 0.4% (zero point four). 

So that's why the result of `iotstack_backup_influxdb` is a just a `.tar`.

> It's a simple `.tar` rather than `.gz.tar` or any other kind of suffix because (a) it's really a mixture of `.tar.gz` *and* other files and (b) because the **only** thing `iotstack_backup_influxdb` does is to create a `.tar`.

Finally, note how those three "mybackup" files are all owned by root. That's because I just called "sudo" (to get permission to read the backup directory properly) without doing anything else. When you use `tar`, you can pre-create the file. Something like:

```
$ touch ~/mysecondbackup.tar
$ sudo tar -cf ~/mysecondbackup.tar -C ~/IOTstack/backups/influxdb/db .
```

The `touch` gets an empty file with the correct ownership (pi:pi) which `tar` is happy to use. I don't believe you can do that with `zip`. If the file exists, it is assumed to actually be a .zip with something in it. In other words, with zip, you usually wind up having to fix the permissions post-hoc.

Hope this helps.

**Comment by Oxmaster** [comment]
Thank you for explaining tar, I used zip because I didn't know the difference, just that it's also just an archive and I have not seen a tar.gz in wild but you can see zip everywhere.

My question sill stands, where does dropbox come in?

My config.yml is
```
backup:
  method: "RCLONE"
  prefix: "dropbox:MyIOTstackBackups"
  retain: 8

restore:
  method: "RCLONE"
  prefix: "dropbox:MyIOTstackBackups"
```

Should I change the `my_influx_backup` script to have `rclone copy $MY_INFLUX_BACKUP dropbox:MyIOTstackBackups` at the end or is there a better method (this seems to work, but I'm not sure if this is how you should be using it)?

**Comment by Paraphraser** [comment]
> Thank you for explaining tar, I used zip because I didn't know the difference, just that it's also just an archive and I have not seen a tar.gz in wild but you can see zip everywhere.

ZIP is more common in the PC and Mac space. On the Mac if I right-click on a folder and choose "compress", a zip is what I get. That said, I see .gz and .tar and .tar.gz all the time on the Mac. The Mac understands them. I can double-click on any of them and they just unpack. The Mac also understands rar and 7z too but I hardly ever see those outside discussions with Windows users.

I suppose it's a *horses for courses* kinda thing.

Unix has always had this ethos of small, simple tools that do one thing and do it well. That's why you often see chained commands with "pipe" between them, like:

```
$ echo "fred" | sudo tee -a someprotected.file
$ ls -1 | wc -l 
```

The idea of "tar" to wrap-up a directory structure and then a compression routine to do something to the output of tar is "natural" to Unix. Although I did things in steps in my earlier reply, and despite `tar` being able to compress/decompress on-the-fly via the `-z` option, the most natural form of archiving on Unix is probably:

```
$ cd test
$ tar -c . | gzip >~/mytest.tar.gz
```

In words:

1. Be in the right directory.
2. run `tar` telling it to `-c` (create) a tar file with `.` (the contents of the current directory). The tar file gets written to standard output so the `|` (pipe) connects that to standard input of `gzip` which writes the compressed result to its standard output, which the shell `>` redirects to the named file.

Anyway, the **purpose** of `iotstack_backup` and the two companion scripts it calls is to create files that can be used by `iotstack_restore` and the two companion scripts that that script calls. If you change `iotstack_backup_influxdb` to create a zip then you also have to change `iotstack_restore_influxdb` to expect a zip. As I pointed out, all you gain from switching to zip is a lot of time to produce very little gain.

<hr>

> My question sill stands, where does dropbox come in?

I interpreted your original question as saying you had that part sorted out, and that the help you needed was in a glue script to connect what you already knew how to do in Dropbox with the result of running `iotstack_backup_influxdb`.

OK. Let's go back a couple of steps.

```
$ tree .config/
.config/
├── iotstack_backup
│   └── config.yml
├── procps
└── rclone
    └── rclone.conf
```

The config.yml file is **only** relevant to `iotstack_backup` and `iotstack_restore`. You aren't using those scripts to that particular file is irrelevant.

To give you a little more detail, the idea of the config.yml is to avoid the user having to tweak the `iotstack_backup` and `iotstack_restore` scripts. The config.yml tells `iotstack_backup` how (protocol) and where (path prefix) to send the results of a backup, and also tells `iotstack_restore` how and where to fetch a backup so that it can be restored.

There's nothing stopping anyone from tweaking any of the scripts. The problem that arises from tweaking is that you can't then just do a `git pull` and get the latest and greatest. You also have to reinvent your customisation wheel each time I change the master scripts on GitHub.

Out of the box, `iotstack_backup` understands RCLONE, RSYNC and SCP but its depth of "understanding" is limited:

* SCP just copies the results of the **current** backup to the target.
* RSYNC synchronises the **folder** containing all backups, including the current backup.
* RCLONE assumes it is synchronising the **folder** containing all backups, including the current backup, with Dropbox.

> Strictly speaking, it can be any target RCLONE understands but there are no guarantees that the command syntax used by `iotstack_backup` when it calls `rclone` will actually work with other kinds of targets. It's a subtle but important caveat.

Because `iotstack_backup` prunes `~/IOTstack/backups` at the end of each run, it follows that the targets of RSYNC and RCLONE will also get pruned - on the **next** run. Conversely, because SCP is only ever copying the **current** backup to the target, the target will never be pruned.

What *you* want to do is a sort of mixture between SCP and RCLONE. You want to send the **current** backup to Dropbox. At least, that's my understanding of your question.

This is where the `rclone.conf` comes in. I'm assuming you have already set that up and that it looks something like:

```
[dropbox]
type = dropbox
token = {"access_token":"YOUR DROPBOX TOKEN","token_type":"bearer","refresh_token":"SOME OTHER TOKEN ","expiry":"SOME TIMESTAMP"}
```

In the above, the (case-sensitive) "dropbox" in "type = dropbox" represents "the access mechanism to reach your top-level folder on Dropbox".

In your `config.yml` you had `dropbox:MyIOTstackBackups` which is interpreted to mean "the folder named `MyIOTstackBackups` at the top-level of Dropbox".

Earlier I said that `iotstack_backup` synchronises the **folder** with Dropbox. If you look at the script, you will see `rclone sync`. All that really needs to happen is to change the `sync` to `copy`.

Rolling all that together:

```
#!/usr/bin/env bash

#define the folder where you want to keep your influx backups
MY_INFLUX_BACKUP_FOLDER="$HOME/my_influx_backups"

# make sure that folder exists
mkdir -p "$MY_INFLUX_BACKUP_FOLDER"

# form the filename that you want
MY_INFLUX_BACKUP_FILE="dbbackup$(date +"%Y-%m-%d").tar"

# form the path
MY_INFLUX_BACKUP="$MY_INFLUX_BACKUP_FOLDER/$MY_INFLUX_BACKUP_FILE"

# run the backup process
iotstack_backup_influxdb "$MY_INFLUX_BACKUP"

# copy the backup to Dropbox
rclone copy "$MY_INFLUX_BACKUP" dropbox:MyIOTstackBackups
```

The question to ask yourself, however, is "what if this fails?"

There's a tight coupling in that rclone which assumes your local network is working, you have a viable path to the Internet, that Dropbox is working, and that everything continues to work while the copy operation is running.

I agree that those assumptions will hold most of the time.

BUT…

You only want to run this once a month, right?

The way `iotstack_backup` works out of the box is to synchronise the folder. That means that a transient comms breakdown today will probably be "fixed" tomorrow and the backups will arrive on Dropbox, eventually. In your case with a `copy`, if that copy doesn't work "first time" then you're sort of cruelled. The only recovery mechanism is to do it by hand.

You could use `iotstack_backup`, just as it comes, and only invoke it once a month.

The alternative would be to sync the folder, rather than the file, by replacing the last two lines of the above script with:

```
# copy the backup to Dropbox
rclone copy "$MY_INFLUX_BACKUP_FOLDER" dropbox:MyIOTstackBackups
```

Then, you can protect yourself against network transients by having a second cron job that runs more frequently (like once a day). Something like:

```
#!/usr/bin/env bash

#define the folder where you want to keep your influx backups
MY_INFLUX_BACKUP_FOLDER="$HOME/my_influx_backups"

# make sure that folder exists
mkdir -p "$MY_INFLUX_BACKUP_FOLDER"

# synchronise with Dropbox
rclone copy "$MY_INFLUX_BACKUP" dropbox:MyIOTstackBackups
```

Most of the time it will do nothing. Until the day when it might save your bacon.
